core/multigrid/pgm.cpp

```diff
@@ -227,10 +237,238 @@ void Pgm<ValueType, IndexType>::generate()
+        generate_coarse(exec, local_matrix.get(), num_agg, agg_);
+
+    return std::tie(prolong_row_gather, coarse_matrix, restrict_sparsity);
+}
+
+
+template <typename IndexType>
+void communicate(
+    std::shared_ptr<const experimental::distributed::MatrixBase<IndexType>>
+        matrix,
+    const array<IndexType>& local_agg, array<IndexType>& non_local_agg)
+{
+    auto exec = gko::as<LinOp>(matrix)->get_executor();
+    const auto comm =
+        gko::as<experimental::distributed::DistributedBase>(matrix)
+            ->get_communicator();
+    auto send_sizes = matrix->get_send_sizes();
+    auto recv_sizes = matrix->get_recv_sizes();
+    auto send_offsets = matrix->get_send_offsets();
+    auto recv_offsets = matrix->get_recv_offsets();
+    auto gather_idxs = matrix->get_gather_idxs();
+    auto total_send_size = send_offsets.back();
+    auto total_recv_size = recv_offsets.back();
+
+    array<IndexType> send_agg(exec, total_send_size);
+    exec->run(pgm::make_gather_index(
+        send_agg.get_num_elems(), local_agg.get_const_data(),
+        gather_idxs.get_const_data(), send_agg.get_data()));
+
+    auto use_host_buffer = experimental::mpi::requires_host_buffer(exec, comm);
+    array<IndexType> host_recv_buffer(exec->get_master());
+    array<IndexType> host_send_buffer(exec->get_master());
+    if (use_host_buffer) {
+        host_recv_buffer.resize_and_reset(total_recv_size);
+        host_send_buffer.resize_and_reset(total_send_size);
+        exec->get_master()->copy_from(exec, total_send_size,
+                                      send_agg.get_data(),
+                                      host_send_buffer.get_data());
+    }
+    auto type = experimental::mpi::type_impl<IndexType>::get_type();
+
+    const auto send_ptr = use_host_buffer ? host_send_buffer.get_const_data()
+                                          : send_agg.get_const_data();
+    auto recv_ptr = use_host_buffer ? host_recv_buffer.get_data()
+                                    : non_local_agg.get_data();
+    exec->synchronize();
+    comm.all_to_all_v(use_host_buffer ? exec->get_master() : exec, send_ptr,
+                      send_sizes.data(), send_offsets.data(), type, recv_ptr,
+                      recv_sizes.data(), recv_offsets.data(), type);
+    if (use_host_buffer) {
+        exec->copy_from(exec->get_master(), total_recv_size, recv_ptr,
+                        non_local_agg.get_data());
+    }
+    return;
+}
+
+
+template <typename IndexType>
+struct larger_index {
+    using type = IndexType;
+};
+
+template <>
+struct larger_index<gko::int32> {
+    using type = gko::int64;
+};
 
-    this->set_multigrid_level(prolong_row_gather, coarse_matrix,
-                              restrict_sparsity);
+
+template <typename ValueType, typename IndexType>
+void Pgm<ValueType, IndexType>::generate()
+{
+    using csr_type = matrix::Csr<ValueType, IndexType>;
+#if GINKGO_BUILD_MPI
+    if (auto matrix = std::dynamic_pointer_cast<
```

Comment:
This implementation is quite lengthy, maybe you can split that up into several sub methods to make the if/else clause easier to read.